### PR TITLE
Updated rules validation on MDU

### PR DESCRIPTION
**Pull request by kritika-singh3**
Original PR: #6555  
 - MDU will go through if there exists at least one material which can refer to the secret config specified
 - MDU will throw `RulesViolationException` if all the pipeline which utilizes the material with the same fingerprint violates the secret config rules.

#### Preview

Screen Shot 2019-06-21 at 3 02 24 PM

### CODE REVIEW

**Comment by kritika-singh3** [comment]
Log snippet:

Screen Shot 2019-06-21 at 3 36 05 PM

File: server/src/main/java/com/thoughtworks/go/server/service/RulesService.java [hunk 1]
```diff
@@ -41,18 +46,103 @@ public RulesService(GoConfigService goConfigService) {
         this.goConfigService = goConfigService;
     }
 
-    public boolean validateSecretConfigReferences(ScmMaterial scmMaterial) {
+    public boolean validateSecretConfigReferences1(ScmMaterial scmMaterial) {
```
Comment (bdpiparva) [inline server/src/main/java/com/thoughtworks/go/server/service/RulesService.java]: Should this be removed?

File: server/src/main/java/com/thoughtworks/go/server/service/RulesService.java [hunk 2]
```diff
@@ -41,18 +46,103 @@ public RulesService(GoConfigService goConfigService) {
...
+    public boolean validateSecretConfigReferences(ScmMaterial scmMaterial) {
+        List<CaseInsensitiveString> pipelines = goConfigService.pipelinesWithMaterial(scmMaterial.getFingerprint());
+
+        HashMap<CaseInsensitiveString, StringBuilder> pipelinesWithErrors = new HashMap<>();
+        pipelines.forEach(pipelineName -> {
+            MaterialConfig materialConfig = goConfigService
+                    .findPipelineByName(pipelineName)
+                    .materialConfigs()
+                    .getByMaterialFingerPrint(scmMaterial.getFingerprint());
+            if (materialConfig instanceof ScmMaterialConfig) {
```
Comment (bdpiparva) [inline server/src/main/java/com/thoughtworks/go/server/service/RulesService.java]: I don't think this is required as the method is only taking ScmMaterial as an input

File: server/src/main/java/com/thoughtworks/go/server/service/RulesService.java [hunk 3]
```diff
@@ -41,18 +46,103 @@ public RulesService(GoConfigService goConfigService) {
...
+                PipelineConfigs group = goConfigService.findGroupByPipeline(pipelineName);
+                ScmMaterialConfig scmMaterialConfig = (ScmMaterialConfig) materialConfig;
+                SecretParams secretParams = SecretParams.parse(scmMaterialConfig.getPassword());
+                secretParams.forEach(secretParam -> {
+                    String secretConfigId = secretParam.getSecretConfigId();
+                    SecretConfig secretConfig = goConfigService.getSecretConfigById(secretConfigId);
+                    if (secretConfig == null) {
+                        addError(pipelinesWithErrors, pipelineName, format("\nPipeline '%s' is referring to none-existent secret config '%s'.", pipelineName, secretConfigId));
```
Comment (bdpiparva) [inline server/src/main/java/com/thoughtworks/go/server/service/RulesService.java]: Do we need  `\n` here?

File: server/src/main/java/com/thoughtworks/go/server/service/RulesService.java [hunk 4]
```diff
@@ -41,18 +46,103 @@ public RulesService(GoConfigService goConfigService) {
...
+                    } else if (!secretConfig.canRefer(group.getClass(), group.getGroup())) {
+                        addError(pipelinesWithErrors, pipelineName, format("\nPipeline '%s' does not have permission to refer to secrets using secret config '%s'", pipelineName, secretConfigId));
```
Comment (bdpiparva) [inline server/src/main/java/com/thoughtworks/go/server/service/RulesService.java]: same as above

File: server/src/main/java/com/thoughtworks/go/server/service/RulesService.java (hunk 3 above)
Comment (kritika-singh3) [reply server/src/main/java/com/thoughtworks/go/server/service/RulesService.java]: No, we don't. We are using `StringUtils.join(list, '\n')`. I will update it.

**Pull request merged**